gcc/rust/resolve/rust-name-resolver-2.0.h

```diff
@@ -0,0 +1,181 @@
+		//         ^
+}
+```
+
+This allows us to revisit scopes previously visited in later phases of the name
+resolution, and add more information if necessary.
+
+  2. `Late`
+
+  `Late` name resolution possesses some unique challenges since Rust's name
+resolution rules are extremely complex - variable shadowing, variable capture in
+closures (but not inner functions!)... You can have a look at a fucked up
+example here:
+
+https://rustc-dev-guide.rust-lang.org/name-resolution.html#scopes-and-ribs
+
+This requires us to think about what exactly to put in our `Scope`s and what to
+do with our `Rib`s - and how it affects our data structures. For example, in the
+above example, `rustc` demonstrates how multiple `Rib`s can be created inside of
+a single lexical scope for variables, as the Rust programming language allows
+shadowing.
+
+    TODO: Mention macro hygiene and that it is the same
+    TODO: How does this affect our data structures?
+    TODO: Last challenge - reuse the same APIs to allow the typechecker to not
+change?
+    TODO: Mention that ForeverStack is templated to make sure that behavior is
+correct
+*/
+
+// Now our resolver, which keeps track of all the `ForeverStack`s we could want
+class Resolver
+{
+public:
+  /**
+   * Insert a new value in the current rib.
+   *
+   * @param name Name of the value to insert.
+   * @param id This value's ID, e.g the function definition's node ID.
+   * @param ns Namespace in which to insert the value.
+   */
+  tl::expected<NodeId, DuplicateNameError> insert (Identifier name, NodeId id,
+						   Namespace ns);
+
+  /**
+   * Run a lambda in a "scoped" context, meaning that a new `Rib` will be pushed
+   * before executing the lambda and then popped. This is useful for all kinds
+   * of scope in the language, such as a block expression or when entering a
+   * function. This variant of the function enters a new scope in *all*
+   * namespaces, while the second variant enters a scope in *one* namespace.
+   *
+   * @param rib New `Rib` to create when entering this scope. A function `Rib`,
+   *        or an item `Rib`... etc
+   * @param scope_id node ID of the scope we are entering, e.g the block's
+   *        `NodeId`.
+   * @param lambda Function to run within that scope
+   * @param path Optional path of the scope. This is useful for scopes which
+   *        affect path resolution, such as modules. Defaults to an empty
+   *        option.
+   */
+  // FIXME: Do we want to handle something in particular for expected within the
+  // scoped lambda?
+  void scoped (Rib rib, NodeId scope_id, std::function<void (void)> lambda,
+	       tl::optional<Identifier> path = {});
+  void scoped (Rib rib, Namespace ns, NodeId scope_id,
+	       std::function<void (void)> lambda,
+	       tl::optional<Identifier> path = {});
+
+  ForeverStack<Namespace::Values> values;
+  ForeverStack<Namespace::Types> types;
+  ForeverStack<Namespace::Macros> macros;
```

Comment:
Should'nt those be private ? I thought the `insert` function would ensure a correct usage.

Reply:
ah, yes they should be private - I had taken them out for debugging purposes

Reply:
after adding more to the Early name resolver I wonder if it doesn't make sense to keep these as public

Reply:
we'll investigate in following PRs, but for now I think we should keep it public